Batchify value_info in resnet50-v1-12 models

jantonguirao opened the pull request:
[A previous PR](https://github.com/onnx/models/pull/537) changed the batch_size in ResNet models in inputs/outputs from `1` to `'N'`. However, the previous PR did not change the batch size in `value_info`, which couple of the models have.

This PR fixes those two models. The changes were applied with a script:
*batchify2.py*
```
import onnx
import sys


#for filename in sys.argv[1:]:

def batchify(filename):
    print("Processing ", filename)
    model = onnx.load(filename)

    in0 = model.graph.input[0]
    assert 'data' in in0.name, f"Name is {in0.name}"

    changed = False
    if in0.type.tensor_type.shape.dim[0].dim_value == 1:
        in0.type.tensor_type.shape.dim[0].dim_param = 'N'
        changed = True
    else:
        assert in0.type.tensor_type.shape.dim[0].dim_param == 'N'

    out0 = model.graph.output[0]
    assert (len(model.graph.output) == 1)

    if out0.type.tensor_type.shape.dim[0].dim_value == 1:
        out0.type.tensor_type.shape.dim[0].dim_param = 'N'
        changed = True
    else:
        assert out0.type.tensor_type.shape.dim[0].dim_param == 'N'

    for i in range(len(model.graph.value_info)):
        vi = model.graph.value_info[i]
        if vi.type.tensor_type.shape.dim[0].dim_value == 1:
            vi.type.tensor_type.shape.dim[0].dim_param = 'N'
            changed = True
        else:
            assert vi.type.tensor_type.shape.dim[0].dim_param == 'N'

    onnx.save_model(model, filename)

    new_model = onnx.load_model(filename)
    new_model.graph.input[0].type.tensor_type.shape.dim[0].dim_param == 'N'
    new_model.graph.output[0].type.tensor_type.shape.dim[0].dim_param == 'N'

    for i in range(len(model.graph.value_info)):
        assert new_model.graph.value_info[i].type.tensor_type.shape.dim[0].dim_param == 'N'
        print(new_model.graph.value_info[i].name, " : ", new_model.graph.value_info[i].type.tensor_type.shape)
    print(in0)
    print(out0)
    print("#"*50)
    return changed

if __name__ == "__main__":
    changed_any = False
    for filename in sys.argv[1:]:
        print(f"Processing {filename}")
        changed_any = changed_any or batchify(filename)
    if not changed_any:
        sys.exit(1)
```
And run like this:
```
cd ~/git/models/vision/classification/resnet/model
git lfs pull --include="resnet*" --exclude=""

for onnxfile in *.onnx; do
    python ~/batchify2.py $onnxfile
done

for tarfile in *.tar.gz; do
    tar xzvf $tarfile
    python ~/batchify2.py */*onnx && rm $tarfile && tar cvzf $tarfile */
    rm -rf */
done
```

jcwchen reviewed (approved):
LGTM. Thank you for the quick fix!

IzanCatalan commented:
Hello @jantonguirao @jcwchen , very useful this python script. It works fine in vgg16 onnx models too. However I haver two problems:

1) With Mobilenet onnx models I get an error:
Traceback (most recent call last):
  File "batchify.py", line 56, in <module>
    changed_any = changed_any or batchify(filename)
  File "batchify.py", line 12, in batchify
    assert 'data' in in0.name, f"Name is {in0.name}"
AssertionError: Name is input

2) Moreover, with Densenet  I get also another error:
Traceback (most recent call last):
  File "batchify.py", line 56, in <module>
    changed_any = changed_any or batchify(filename)
  File "batchify.py", line 36, in batchify
    assert vi.type.tensor_type.shape.dim[0].dim_param == 'N'
AssertionError

I would like to know why this happens and what is the reason for those asserts and if there is a solution?

Thanks!

jantonguirao commented:
@IzanCatalan There's no special reason for the asserts. When I was writing the script for ResNet, I noticed that the input to the network was called `data`, so I added that assert to make sure that I was not batchifying anything else. I think you can safely remove that one. 

The other one is there as a sanity check:
```
        if vi.type.tensor_type.shape.dim[0].dim_value == 1:
            vi.type.tensor_type.shape.dim[0].dim_param = 'N'
            changed = True
        else:
            assert vi.type.tensor_type.shape.dim[0].dim_param == 'N'
```
I knew that ResNet models had a hardcoded `1` batched dimension, so I was making sure that if we encounter such dimension, we change it to `N` and if there's anything else, it must be already `N`. Now, I am not familiar with the model that you are trying to batchify, but it might be that it has other value infos that you don't want to batchify.

Generally speaking, I wrote this script for the task of adding a batch dimension to ResNet models. It might be reused with other models, but probably needs some careful examination of the model first.
